src/CodeGen_LLVM.cpp

```diff
@@ -2457,6 +2459,150 @@ void CodeGen_LLVM::codegen_predicated_vector_load(const Load *op) {
+    Expr val_expr = op->value;
+    Halide::Type value_type = op->value.type();
+    // For atomicAdd, we check if op->value - store[index] is independent of store.
+    // For llvm version < 9, the atomicRMW operations only support integers so we also check that.
+    Expr equiv_load = Load::make(value_type, op->name,
+                                 op->index,
+                                 Buffer<>(),
+                                 op->param,
+                                 op->predicate,
+                                 op->alignment);
+    Expr delta = simplify(common_subexpression_elimination(op->value - equiv_load));
+#if LLVM_VERSION >= 90
+    bool is_atomic_add = !expr_uses_var(delta, op->name);
+    if (is_ptx_atomic_store) {
+        if (value_type.is_float() &&
+                value_type.bits() == 64 &&
+                !target.has_feature(Target::CUDACapability61)) {
+            // AtomicRMW complains when the GPU backend doesn't support the 
+            // corresponding atomic add operation.
+            is_atomic_add = false;
+        }
+    }
+#else
+    bool is_atomic_add = value_type.is_int_or_uint() &&
+                         !expr_uses_var(delta, op->name);
+#endif
+    if (is_atomic_add) {
+        Value *val = codegen(delta);
+        if (value_type.is_scalar()) {
+            Value *ptr = codegen_buffer_pointer(op->name,
+                                                op->value.type(),
+                                                op->index);
+            // llvm 9 has FAdd which can be used for atomic floats.
+#if LLVM_VERSION >= 90
+            if (value_type.is_float()) {
+                builder->CreateAtomicRMW(AtomicRMWInst::FAdd, ptr, val, AtomicOrdering::Monotonic);
+            } else {
+                builder->CreateAtomicRMW(AtomicRMWInst::Add, ptr, val, AtomicOrdering::Monotonic);
+            }
+#else
+            builder->CreateAtomicRMW(AtomicRMWInst::Add, ptr, val, AtomicOrdering::Monotonic);
+#endif
+        } else {
+            Value *index = codegen(op->index);
+            // Scalarize vector store.
+            for (int i = 0; i < value_type.lanes(); i++) {
+                Value *lane = ConstantInt::get(i32_t, i);
+                Value *idx = builder->CreateExtractElement(index, lane);
+                Value *v = builder->CreateExtractElement(val, lane);
+                Value *ptr = codegen_buffer_pointer(op->name, value_type.element_of(), idx);
+#if LLVM_VERSION >= 90
+                if (value_type.is_float()) {
+                    builder->CreateAtomicRMW(AtomicRMWInst::FAdd, ptr, v, AtomicOrdering::Monotonic);
+                } else {
+                    builder->CreateAtomicRMW(AtomicRMWInst::Add, ptr, v, AtomicOrdering::Monotonic);
+                }
+#else
+                builder->CreateAtomicRMW(AtomicRMWInst::Add, ptr, v, AtomicOrdering::Monotonic);
+#endif
+            }
+        }
+    } else {
+        // We want to create the following CAS loop:
```

Comment:
Can this logic be moved into a runtime function, so that it can be written in C or .ll? Codegen_LLVM is already huge.

Reply:
Hmm might be difficult since we want to inline the value computation inside the loop...

Reply:
Fair enough.